pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java

```diff
@@ -100,7 +102,7 @@ protected void grabCnx() {
                             log.error("[{}] [{}] Unexpected exception after the connection",
                                     state.topic, state.getHandlerName(), e);
                         }
-                        duringConnect = false;
+                        duringConnect.compareAndSet(true, false);
```

Comment:
It's right.

Reply:
I do not think so, this will makes more than one `granCnx` running at the same time. For example:

| time | `task 1` | `task 2` | `task 3` |
| --- | --- | --- | --- |
| 1 | grab connection | 
| 2 | `duringConnect` --> `true` | 
| 3 | recconect | 
| 4 | `duringConnect` --> `false` |
| 5 | | grab connection |
| 6 | | `duringConnect` --> `true` |
| 7 | after the future complete |
| 8 | `duringConnect` --> `false` |
| 9 | | | grab connection |
| 10 | | | `duringConnect` --> `true` |

Suggestion to use `compareAndSet` everywhere.

Reply:
@poorbarcode do you mean check the return value of `compareAndSet` and return if it's `false`?

Reply:
> Suggestion to use compareAndSet everywhere.

It does not resolve anything. CAS on a boolean variable is meaningless. `x.compareAndSet(true, false)` is equivalent to `x.set(false)`. What matters is the returned value of CAS, which can check if `x` was true before setting it with false. @poorbarcode

Reply:
I think what affects is  the step 7 and 8. The root cause is caused by https://github.com/apache/pulsar/pull/20595#discussion_r1245264041, when `reconnectLater` is called in `connectionOpened`, since the future is completed exceptionally, `handleConnectionError` will be called and `reconnectLater` will be called again.

```java
            cnxFuture.thenCompose(cnx -> connection.connectionOpened(cnx))
                    .thenAccept(__ -> duringConnect.set(false))
                    .exceptionally(this::handleConnectionError);
```

Reply:
I think this issue has been solved

Reply:
It's hard to simulate this race condition. I only added tests to verify `duringConnect.set(false)` and `duringConnect.compareAndSet(false, true)` are both called once each time.